test/Semantics/data01.f90

```diff
@@ -37,12 +37,25 @@ subroutine CheckRepeat
   DATA myName%age / digits(myAge) * 35 /
 end
 
-!subroutine CheckValue
-!  use m1
-!  !C883
-!  !ERROR: Derived type 'persn' not found
-!  DATA myname / persn(2, 'Abcd Efgh') /
-!  !C884
-!  !ERROR: Structure constructor in data value must be a constant expression
-!  DATA myname / person(myAge, 'Abcd Ijkl') /
-!end
+subroutine CheckValue
+  use m1
+  !OK: constant structure constructor
+  data myname / person(1, 'Abcd Ijkl') /
+  !C883
+  !ERROR: Must have INTEGER type, but is CHARACTER(1)
+  data myname / persn(2, 'Abcd Efgh') /
+  !C884
```

Comment:
@tskeith, Is there a change in error message emitted? This does not seem to be a valid error message to me. It was emitting something like "Derived type 'persn' not found".

Reply:
The problem is that the compiler can't tell if `persn` was meant to be an array or a derived type. It now parses it as an array element reference and changes it to a derived type constructor if it is the name of a known derived type. Since it's not, it leaves it as an array element reference, so the error is that a character can't be used as an array index.

We might be able to improve the error message by reporting that it neither an array nor a derived type.

Reply:
I find that as a better way if possible. That can cover the cases where the constant part of any ConstantSubobject is not defined. If that requires any extra implementation in Semantic part, I can take a look at that.